### PR TITLE
feat(casting): support ignored_ops in cast_fp32_to_fp16 and cast_to_16_bit_precision (#7)

### LINKED ISSUE #7: FP16 casting pass does not guard against activation-level overflow (softplus, exp, logsumexp)
## Problem

The FP16 casting pass in `casting.py` correctly guards against **static tensor overflow** (weights/constants > FP16_MAX = 65504), but does not account for **activation-level overflow** — where operations like `exp(x)` produce intermediate values > 65504 at runtime even though the inputs are within fp16 range.

### Example: softplus

A model with a `Softplus()` activation layer:
1. `cast_fp32_to_fp16()` successfully converts all weights and inputs to fp16 ✅
2. The converter decomposes `softplus(x) = log(1 + exp(x))`
3. At runtime on Apple Neural Engine, when `x > 10.4`:
   - `exp(10.4)` ≈ 32,900 (fits fp16)
   - `exp(11.0)` ≈ 59,874 (barely fits fp16)
   - `exp(11.1)` ≈ 66,686 → **OVERFLOW** → output collapses to 0

The current `check_tensor_overflow_fp16()` and `handle_overflow_op()` logic only checks scalar/tensor *values*, not whether the *computation graph* will produce intermediate overflows.

### Affected Operations

| Operation | Naive Form | fp16 Overflow Threshold |
|-----------|-----------|----------------------|
| softplus | `exp(x)` | x ≈ 10.4 |
| logsumexp | `sum(exp(x_i))` | x ≈ 7.63 |
| logcumsumexp | `cumsum(exp(x_i))` | x ≈ 11.09 |

### The Compound Effect

When `coreai-optimization` applies weight compression (palettization, quantization) AND fp16 casting together:
1. Quantization introduces rounding errors in weights
2. These errors can shift activation distributions
3. Values that were safely below the overflow threshold may now exceed it
4. The casting pass has no mechanism to detect or prevent this

### Proposed Fix

Add an `activation_overflow_audit` pass that:
1. Identifies ops in the graph whose intermediates can overflow fp16 (`exp`, `log(1+exp(...))`)
2. Flags them for stable decomposition or fp32 accumulation
3. Integrates with the existing `handle_overflow_op` / `handle_non_overflow_op` classification

### Prior Art

- `apple/coremltools` PRs #2725, #2726, #2727 fix the converter-level decomposition
- `apple/coreai-torch` PR #22 adds stable converters for softplus/mish/logsumexp
- This issue addresses the *optimization* layer, where the interaction between quantization and fp16 creates compound failures

### Environment

- coreai-optimization: latest (cloned June 21, 2026)
- Related: apple/coreai-torch#21, apple/coremltools#2687

### PR DESCRIPTION
**Pull request by rohith500**
### Summary

This PR adds an `ignored_ops` parameter to `cast_fp32_to_fp16` and `cast_to_16_bit_precision`, allowing users to specify operations that should be preserved in `torch.float32` during 16-bit casting. This resolves #7 by providing a deterministic mechanism to prevent intermediate activation-level overflow in operations such as `exp`, `logsumexp`, `softplus`, or `mish`.

### Motivation & Background

As discussed in #7, IEEE 754 half-precision float (`FP16`) has a maximum finite limit of `65,504`. Neural network activation functions involving exponentials (such as $\text{Softplus}(x) = \ln(1 + \exp(x))$, $\text{LogSumExp}$, or $\text{Mish}$) exceed this limit whenever intermediate activations $x > 11.0$ (e.g. $\exp(15.0) \approx 3.27 \times 10^6$), causing the operation to evaluate to $+\infty$. On hardware backends like Apple Silicon / ANE, this overflow causes numerical collapse downstream into `0.0` or `NaN`.

Because `cast_fp32_to_fp16` performs graph transformations prior to runtime without inspecting dynamic activation magnitudes, it eagerly converts these compute ops to `torch.float16`. Following maintainer guidance from @crowbat in #7 (Direction 2: *"allowing the user to specify specific ops or op types to ignore being casted"*), this PR enables users to selectively skip casting for sensitive operations.

### Changes

1. **Selective Op Identification (`is_ignored_op`)**:
   - Added `is_ignored_op` in `coreai_opt._utils.casting_utils`.
   - Supports 5 flexible identifier formats:
     - `torch._ops.OpOverload` (e.g. `torch.ops.aten.exp.default`)
     - `torch._ops.OpOverloadPacket` (e.g. `torch.ops.aten.exp`)
     - Standard PyTorch callables (e.g. `torch.exp`)
     - String op names (e.g. `"exp"`, `"aten.exp"`, `"aten::exp"`)
     - String node names (e.g. `"exp_1"`, targeting specific graph nodes)

2. **Casting Engine Integration (`_FP16Casting`)**:
   - In `_FP16Casting.iterate_nodes_and_insert_casts`, checks `is_ignored_op(node, self._ignored_ops)` before creation and general compute ops.
   - Routes ignored ops directly to `handle_overflow_op(node)`, which preserves the op in `torch.float32`, inserts `FP16 -> FP32` upcasts on any FP16 inputs, and inserts `FP32 -> FP16` downcasts on outputs for downstream consumers.
   - Redundant intermediate casts between consecutive ignored ops (e.g. `exp` $\to$ `log1p`) are automatically collapsed by the existing `_cleanup_casts` engine into a direct FP32 chain with zero overhead.

3. **Public API Updates**:
   - Added `ignored_ops: Sequence[Callable | str] | None = None` to `cast_fp32_to_fp16` and `cast_to_16_bit_precision`.
   - Fully backward-compatible; defaults to `None`.

4. **Testing & Documentation**:
   - Added `TestIsIgnoredOp` unit tests in `tests/casting/test_casting_utils.py` covering all matching styles and edge cases.
   - Added `TestSelectiveOpSkipping` end-to-end tests in `tests/casting/test_casting.py` verifying activation overflow prevention at $x=15.0$, all matching types, specific node targeting, consecutive op cast elimination, creation op handling, and top-level `cast_to_16_bit_precision` forwarding.
   - Updated documentation in `docs/src/utils/casting.md`.
   - Added Towncrier changelog fragment `changelog.d/7.added`.

### Verification

- All unit and end-to-end tests pass (106/106 in `tests/casting/`):
  ```bash
  pytest tests/casting/ -v
  ```
- All repository quality checks pass (26/26):
  ```bash
  make check
  ```

Closes #7

### CODE REVIEW

File: src/coreai_opt/_utils/casting_utils.py [hunk 1]
```diff
@@ -626,6 +628,60 @@ def _check(arg: object) -> tuple[bool, bool]:
     return _check((*node.args, *node.kwargs.values()))
 
 
+def is_ignored_op(
```
Comment (crowbat) [inline src/coreai_opt/_utils/casting_utils.py]: In trying to allow such flexibility for users to match on so many different criteria (name, op type name, torch function name), I feel like the matching can get brittle. In theory, node names are mutable and in some weird corner case might end up being the same as an op type for which the user did not intend to match with for example.

Can we simplify the supported inputs to either
1) A set of OpOverload or OpOverloadPacket op instances (e.g. torch.ops.aten.exp, torch.ops.aten.exp.default, etc.)
2) A predicate function which takes a graph node as input and outputs True or False

In case 1), I'd like to avoid dealing with any kind of string, instead using the actual op overload/packet instances for robust matching. We can operate on the op overload/packet granularity since each graph node's target maps unambiguously to an opoverload/packet, whereas torch callables may not correspond directly to a node (ex. functions like `torch.nn.functional.mish` may get decomposed into intermediate ops). Specifying `torch.nn.functional.mish` in the current state might end up never matching with any ops and it wouldn't be very straightforward in determining why.

Case 2) would handle any more complex use cases by allowing the user to define their own matching function to ignore specific ops, in case option 1) is too broad for them. Here they would still be able to choose to use node names or any other kind of string matching if they wanted to, but the complexity and responsibility for correctly handling edge cases then falls to the user instead of `casting_utils.py`.

**Review by crowbat — COMMENTED** [review]
@rohith500 , thanks for your PR. I have one comment on the proposed node matching.

File: src/coreai_opt/_utils/casting_utils.py (hunk 1 above)
Comment (rohith500) [reply src/coreai_opt/_utils/casting_utils.py]: Hi @crowbat,

Thank you for this thoughtful feedback. That is a great point regarding node name mutability and the decomposition trap for high-level callables like `torch.nn.functional.mish`.

I have updated the PR to adopt this simplified, robust two-pronged approach:

1. **Option 1 (`OpOverload` / `OpOverloadPacket` instances)**:
   - Users can pass a set/collection of `OpOverload` or `OpOverloadPacket` instances (e.g. `{torch.ops.aten.exp}`, `{torch.ops.aten.exp.default}`), or a single op instance.
   - Matching is direct object equality and packet equality (`target in ignored_ops or getattr(target, "overloadpacket", None) in ignored_ops`) with $O(1)$ set lookup.
   - Removed all string parsing, regex, and callable name heuristics.

2. **Option 2 (Predicate function `Callable[[torch.fx.Node], bool]`)**:
   - Users can provide a custom predicate function (e.g. `lambda node: node.name == "exp_1"`, or filtering on `node.meta["nn_module_stack"]`) for fine-grained or contextual control.
   - Any complexity or edge-case handling for node names/scopes is cleanly encapsulated in the user's predicate rather than internal library heuristics.

I also updated `_FP16Casting.__init__` to normalize collection inputs into a set once, preserving $O(1)$ lookup per node during graph traversal.

All 108 tests in `tests/casting/` and all 26 checks in `make check` pass cleanly, and the branch is rebased onto latest `main`. Ready for your re-review!

**Review by crowbat — APPROVED** [review]
Thanks for the changes @rohith500 , looks good to me.
